storagestate/storagestate.go

```diff
@@ -0,0 +1,709 @@
+	}
+
+	storageState.stateChangedChannel <- StateChangedInfo{
+		InstanceID: instanceID,
+		Checksum:   checksum,
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) pushNewStateMessage(
+	instanceIdent cloudprotocol.InstanceIdent, checksum, stateData string,
+) (err error) {
+	if len(storageState.newStateChannel) >= cap(storageState.newStateChannel) {
+		return aoserrors.New("new state channel is full")
+	}
+
+	storageState.newStateChannel <- cloudprotocol.NewState{
+		InstanceIdent: instanceIdent,
+		Checksum:      checksum,
+		State:         stateData,
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) pushStateRequestMessage(
+	instanceIdent cloudprotocol.InstanceIdent, defaultState bool,
+) (err error) {
+	if len(storageState.stateRequestChannel) >= cap(storageState.stateRequestChannel) {
+		return aoserrors.New("state request channel is full")
+	}
+
+	storageState.stateRequestChannel <- cloudprotocol.StateRequest{
+		InstanceIdent: instanceIdent,
+		Default:       defaultState,
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) processWatcher() {
+	for {
+		select {
+		case event, ok := <-storageState.watcher.Events:
+			if !ok {
+				return
+			}
+
+			storageState.Lock()
+
+			for instanceIdent, state := range storageState.statesMap {
+				if state.stateFilePath == event.Name {
+					log.WithFields(log.Fields{
+						"file":       event.Name,
+						"instanceID": state.instanceID,
+					}).Debug("State file changed")
+
+					if state.changeTimer == nil {
+						state.changeTimer = time.NewTimer(StateChangeTimeout)
+
+						go func() {
+							select {
+							case <-state.changeTimer.C:
+								storageState.stateChanged(event.Name, instanceIdent, state)
+
+							case <-state.changeTimerChannel:
+								log.WithField("instanceID", state.instanceID).Debug("Send state change cancel")
+							}
+						}()
+					} else {
+						state.changeTimer.Reset(StateChangeTimeout)
+					}
+
+					break
+				}
+			}
+
+			storageState.Unlock()
+
+		case err, ok := <-storageState.watcher.Errors:
+			if !ok {
+				return
+			}
+
+			log.Errorf("FS watcher error: %s", err)
+		}
+	}
+}
+
+func (storageState *StorageState) stateChanged(
+	fileName string, instanceIdent cloudprotocol.InstanceIdent, state *stateParams,
+) {
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	state.changeTimer = nil
+
+	stateData, checksum, err := getFileDataChecksum(fileName)
+	if err != nil {
+		log.Errorf("Can't get state and checksum: %s", err)
+
+		return
+	}
+
+	// This check is necessary because in UpdateState the data is saved to the file
+	// and at the same time the file change event occurs
+	if bytes.Equal(checksum, state.checksum) {
+		return
+	}
+
+	state.checksum = checksum
+
+	if err := storageState.pushNewStateMessage(
+		instanceIdent, hex.EncodeToString(checksum), string(stateData)); err != nil {
+		log.Errorf("Can't send new state message: %s", err)
+	}
+}
+
+func (storageState *StorageState) setup(
```

Comment:
Could it be renamed to `prepareStorage` similar to `prepareState`?
and `retFolderPath` -> `storagePath`

Reply:
Agree. Rename